File: Source/Game/Items/Relics/PlayerUpgrades.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;

public class PlayerUpgrades
{
    private player Player;
    private Dictionary<Type, int> UpgradeLevels { get; set; } = new Dictionary<Type, int>();

    public PlayerUpgrades(player player)
    {
        Player = player;
        Globals.PLAYER_UPGRADES[player] = this;
    }

    public static void Initialize()
    {
        foreach (var player in Globals.ALL_PLAYERS)
            new PlayerUpgrades(player);
    }

    public static PlayerUpgrades GetPlayerUpgrades(player player)
    {
        if (Globals.PLAYER_UPGRADES.ContainsKey(player))
            return Globals.PLAYER_UPGRADES[player];
        else
            return new PlayerUpgrades(player);
    }
    public void SetUpgradeLevel(Type relicType, int level) => UpgradeLevels[relicType] = level;

    public static void IncreaseUpgradeLevel(Type relicType, unit Unit)
    {
        var player = Unit.Owner;
        GetPlayerUpgrades(player).SetUpgradeLevel(relicType, GetPlayerUpgrades(player).GetUpgradeLevel(relicType) + 1);
    }

    public int GetUpgradeLevel(Type relicType) => UpgradeLevels.TryGetValue(relicType, out var level) ? level : 0;
}

File: src/Game/Items/Relics/PlayerUpgrades.ts
import { Globals } from 'src/Global/Globals'
import { MapPlayer, Unit } from 'w3ts'

export class PlayerUpgrades {
    private Player: MapPlayer
    private UpgradeLevels: Map<string, number> = new Map()

    public constructor(player: MapPlayer) {
        this.Player = player
        Globals.PLAYER_UPGRADES.set(player, this)
    }

    public static Initialize = () => {
        for (const player of Globals.ALL_PLAYERS) new PlayerUpgrades(player)
    }

    public static GetPlayerUpgrades(player: MapPlayer): PlayerUpgrades {
        if (Globals.PLAYER_UPGRADES.has(player)) return Globals.PLAYER_UPGRADES.get(player)!
        else return new PlayerUpgrades(player)
    }

    public SetUpgradeLevel(relicType: string, level: number) {
        return this.UpgradeLevels.set(relicType, level)
    }

    public static IncreaseUpgradeLevel(relicType: string, Unit: Unit) {
        const player = Unit.owner
        PlayerUpgrades.GetPlayerUpgrades(player).SetUpgradeLevel(
            relicType,
            PlayerUpgrades.GetPlayerUpgrades(player).GetUpgradeLevel(relicType) + 1
        )
    }

    public GetUpgradeLevel(relicType: string) {
        return this.UpgradeLevels.get(relicType) || 0
    }
}
